buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2028 @@
+    }
+
+    @Override
+    protected RuntimeException createResourceClosedException() {
+        return bufferIsClosed(this);
+    }
+
+    @Override
+    public CompositeBuffer order(ByteOrder order) {
+        if (this.order != order) {
+            this.order = order;
+            for (Buffer buf : bufs) {
+                buf.order(order);
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public ByteOrder order() {
+        return order;
+    }
+
+    @Override
+    public int capacity() {
+        return capacity;
+    }
+
+    @Override
+    public int readerOffset() {
+        return roff;
+    }
+
+    @Override
+    public CompositeBuffer readerOffset(int index) {
+        prepRead(index, 0);
+        int indexLeft = index;
+        for (Buffer buf : bufs) {
+            buf.readerOffset(Math.min(indexLeft, buf.capacity()));
+            indexLeft = Math.max(0, indexLeft - buf.capacity());
+        }
+        roff = index;
+        return this;
+    }
+
+    @Override
+    public int writerOffset() {
+        return woff;
+    }
+
+    @Override
+    public CompositeBuffer writerOffset(int index) {
+        checkWriteBounds(index, 0);
+        int indexLeft = index;
+        for (Buffer buf : bufs) {
+            buf.writerOffset(Math.min(indexLeft, buf.capacity()));
+            indexLeft = Math.max(0, indexLeft - buf.capacity());
+        }
+        woff = index;
+        return this;
+    }
+
+    @Override
+    public CompositeBuffer fill(byte value) {
+        for (Buffer buf : bufs) {
+            buf.fill(value);
+        }
+        return this;
+    }
+
+    @Override
+    public long nativeAddress() {
+        return 0;
+    }
+
+    @Override
+    public CompositeBuffer makeReadOnly() {
+        for (Buffer buf : bufs) {
+            buf.makeReadOnly();
+        }
+        readOnly = true;
+        return this;
+    }
+
+    @Override
+    public boolean readOnly() {
+        return readOnly;
+    }
+
+    @Override
+    public CompositeBuffer copy(int offset, int length) {
+        checkWriteBounds(offset, length);
+        if (offset < 0 || length < 0) {
+            throw new IllegalArgumentException(
+                    "Offset and length cannot be negative, but offset was " +
+                    offset + ", and length was " + length + '.');
+        }
+        Buffer choice = (Buffer) chooseBuffer(offset, 0);
+        Buffer[] copies;
+
+        if (length > 0) {
+            copies = new Buffer[bufs.length];
+            int off = subOffset;
+            int cap = length;
+            int i;
+            for (i = searchOffsets(offset); cap > 0; i++) {
+                var buf = bufs[i];
+                int avail = buf.capacity() - off;
+                copies[i] = buf.copy(off, Math.min(cap, avail));
+                cap -= avail;
+                off = 0;
+            }
+            copies = Arrays.copyOf(copies, i);
```

Comment:
Maybe this is obvious, but how does this avoid the beginning of the array containing nulls?  The for loop looks like it searches for the first buffer that meets the offset, but starts writing into the middle of `copies`.    Should the loop have a separate reader and writer index for bufs and copies?

Reply:
Good catch!